### PR TITLE
Add prefix search functionality to the search store

### PR DESCRIPTION
**Pull request by scottming**
***Description:*** 

This commit adds the `prefix` function to the `Lexical.RemoteControl.Search.Store` module, allowing for searching entries by prefix. It also implements the `find_by_prefix` callback in the `Lexical.RemoteControl.Search.Store.Backends.Ets` module.

***Changes:***

- Added `prefix` function to `Lexical.RemoteControl.Search.Store`
- Implemented `find_by_prefix` callback in `Lexical.RemoteControl.Search.Store.Backends.Ets`

### CODE REVIEW

File: apps/remote_control/lib/lexical/remote_control/search/store/backends/ets/schemas/v2.ex [hunk 1]
```diff
@@ -30,7 +30,7 @@ defmodule Lexical.RemoteControl.Search.Store.Backends.Ets.Schemas.V2 do
   def to_rows(%Entry{} = entry) do
     subject_key =
       by_subject(
-        subject: to_subject(entry.subject),
+        subject: to_subject_charlist(entry.subject),
```
Comment (scottming) [inline apps/remote_control/lib/lexical/remote_control/search/store/backends/ets/schemas/v2.ex:33]: Do I need to change the key here from by_subject to by_subject_prefix?

File: apps/remote_control/lib/lexical/remote_control/search/store/backends/ets/schemas/v2.ex [hunk 2]
```diff
@@ -62,8 +62,8 @@ defmodule Lexical.RemoteControl.Search.Store.Backends.Ets.Schemas.V2 do
...
   end
 
-  defp to_subject(binary) when is_binary(binary), do: binary
-  defp to_subject(:_), do: :_
-  defp to_subject(atom) when is_atom(atom), do: inspect(atom)
-  defp to_subject(other), do: to_string(other)
+  def to_subject_charlist(charlist) when is_list(charlist), do: charlist
+  def to_subject_charlist(:_), do: :_
+  def to_subject_charlist(atom) when is_atom(atom), do: atom |> inspect() |> to_charlist()
+  def to_subject_charlist(other), do: to_charlist(other)
```
Comment (scohen) [inline apps/remote_control/lib/lexical/remote_control/search/store/backends/ets/schemas/v2.ex:68]: why the name change here?

Comment (scottming) [reply apps/remote_control/lib/lexical/remote_control/search/store/backends/ets/schemas/v2.ex]: Ok, reverted

**Comment by scohen** [comment]
What about all the places we use subject?

File: apps/remote_control/lib/lexical/remote_control/search/store/backends/ets/state.ex [hunk 3]
```diff
@@ -97,6 +98,28 @@ defmodule Lexical.RemoteControl.Search.Store.Backends.Ets.State do
...
+    |> Enum.flat_map(fn {_, id_keys} -> id_keys end)
+    |> MapSet.new()
+    |> Enum.flat_map(&:ets.lookup_element(state.table_name, &1, 2))
+  end
+
+  @dialyzer {:nowarn_function, to_prefix: 1}
+
+  defp to_prefix(prefix) when is_binary(prefix) do
+    [last_char | others] = prefix |> String.to_charlist() |> Enum.reverse()
+    others |> Enum.reverse() |> Enum.concat([last_char | :_])
```
[the remark this answers is not in the document]
Comment (scottming) [reply apps/remote_control/lib/lexical/remote_control/search/store/backends/ets/state.ex:120]: no, what we really need is an improper list.
```elixir
iex(1)> prefix = "abc"
"abc"
iex(2)>     [last_char | others] = prefix |> String.to_charlist() |> Enum.reverse()
~c"cba"
iex(3)>     others |> Enum.reverse() |> Enum.concat([last_char | :_])
[97, 98, 99 | :_]
iex(4)> 'abc' ++ [:_]
[97, 98, 99, :_]
```

**Comment by scottming** [comment]
> What about all the places we use subject?

@scohen I'm not sure if I've understood your question correctly, but regarding how to handle the places where 'subject' was previously used, we continue to use `find_by_subject` to query a specific subject.

File: apps/remote_control/lib/lexical/remote_control/search/store/backends/ets/state.ex (hunk 3 above)
[the remark this answers is not in the document]
Comment (scottming) [reply apps/remote_control/lib/lexical/remote_control/search/store/backends/ets/state.ex:122]: how about change the code like this:

```elixir
{last_char, others} = List.pop_at('abc', -1)
others ++ [last_char | :_]
```
?

[the remark this answers is not in the document]
Comment (scohen) [reply apps/remote_control/lib/lexical/remote_control/search/store/backends/ets/state.ex:122]: `Lists.insert_at(list, -1, :_)`

**Comment by scohen** [comment]
I think i was a little nervous, your're not actually changing the subject that resides in the Entry to a list, just the key in the ETS table.

I think this will require a reindex, correct?

**Comment by scottming** [comment]
> I think this will require a reindex, correct?

yes, Yes, if we don't do a reindex, then references will be unavailable. How can we enforce this?

File: apps/remote_control/lib/lexical/remote_control/search/store/backends/ets/state.ex (hunk 3 above)
[the remark this answers is not in the document]
Comment (scottming) [reply apps/remote_control/lib/lexical/remote_control/search/store/backends/ets/state.ex]: added

**Comment by scohen** [comment]
We'll need to divorce migrations from schemas, or create a new schema that's just like the old one, but with a migration

**Comment by scottming** [comment]
> We'll need to divorce migrations from schemas, or create a new schema that's just like the old one, but with a migration

@scohen To be honest, I'm not sure how to properly achieve this. Adding a new version of the Schema and defining some migration logic in it could achieve this purpose, but it seems like this would duplicate a lot of the code in v2. Is this what you want?

**Review by scohen — APPROVED** [review]

**Pull request merged**